File: lab4/csharp/solution2/Table.cs
using System.Threading;

public class Table
{
    private readonly SemaphoreSlim[] forks = new SemaphoreSlim[5];
    private readonly SemaphoreSlim waiter = new SemaphoreSlim(2, 2); // max 2 одночасно

    public Table()
    {
        for (int i = 0; i < forks.Length; i++)
            forks[i] = new SemaphoreSlim(1, 1);
    }

    public void GetFork(int right, int left)
    {
        waiter.Wait();
        forks[right].Wait();
        forks[left].Wait();
    }

    public void PutFork(int right, int left)
    {
        forks[left].Release();
        forks[right].Release();
        waiter.Release();
    }
}

File: lab4/philosophers2025/solution2/src/Table.java
import java.util.concurrent.Semaphore;

public class Table {
    private final Semaphore[] forks = new Semaphore[5];
    private final Semaphore waiter = new Semaphore(2);

    public Table() {
        for (int i = 0; i < forks.length; i++) {
            forks[i] = new Semaphore(1);
        }
    }

    public void getFork(int right, int left) throws InterruptedException {
        waiter.acquire();
        forks[right].acquire();
        forks[left].acquire();
    }

    public void putFork(int right, int left) {
        forks[left].release();
        forks[right].release();
        waiter.release();
    }
}